src/java.base/share/classes/java/time/format/DateTimeTextProvider.java

```diff
@@ -312,8 +312,10 @@ private Object findStore(TemporalField field, Locale locale) {
         Object store = CACHE.get(key);
         if (store == null) {
             store = createStore(field, locale);
-            CACHE.putIfAbsent(key, store);
-            store = CACHE.get(key);
+            Object prev = CACHE.putIfAbsent(key, store);
+            if (prev != null) {
+                store = prev;
+            }
         }
```

Comment:
You could do better here and use `computeIfAbsent` with `createStore` as its lambda. You could even change the signature of `createStore` to take `Entry<TemporalField, Locale>` as its parameter and then you could have this method be:
```java
    private Object findStore(TemporalField field, Locale locale) {
        return CACHE.computeIfAbsent(createEntry(field, locale), this::createStore);
    }

    ...

    private Object createStore(Entry<TemporalField, Locale> entry) {
        ...
    }
```
This applies to most other changes you made, the one in `ZoneOffset` is the only one that's slightly different because there you have adjustment of `ID_CACHE` too.

Reply:
This behaves slightly different from the old initialization; the concurrent hash map blocks when the mapping function is run, just in case if non-blocking instantiation is what we want. If that's not a problem, I would prefer szegedi's suggestion.

Reply:
@liach advance apologies for nitpicking: `ConcurrentHashMap` doesn't in general block while the mapping function runs. It can block _some_ concurrent updates, namely those that [hash to the same bin](https://github.com/openjdk/jdk/blob/0f801fe6fd2fcc181121f9846f6869ca3a03e18a/src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java#L324) and it won't block concurrent reads. I'm not saying the concern you raised isn't valid, just wanted to clarify that the situation is not nearly as grave as overall blocking; after all it ain't a `Collections.synchronizedMap` :-)